File: backend/app/schemas/research.py
"""
Pydantic schemas for Research API responses.
"""

from typing import Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field


class IVTResponse(BaseModel):
    """Intrinsic Value Target response schema."""
    ticker: str
    intrinsic_value: float = Field(..., description="Calculated intrinsic value per share")
    current_price: float = Field(..., description="Current stock price")
    discount_premium_pct: float = Field(..., description="Discount/premium percentage (positive = discount)")
    wacc: float = Field(..., description="Weighted Average Cost of Capital")
    terminal_growth_rate: float = Field(..., description="Terminal growth rate assumption")
    calculated_at: datetime = Field(..., description="When IVT was calculated")

    class Config:
        json_schema_extra = {
            "example": {
                "ticker": "AAPL",
                "intrinsic_value": 185.50,
                "current_price": 175.20,
                "discount_premium_pct": 5.88,
                "wacc": 0.085,
                "terminal_growth_rate": 0.025,
                "calculated_at": "2024-01-15T10:30:00Z"
            }
        }


class RiskLevelResponse(BaseModel):
    """Risk Level assessment response schema."""
    ticker: str
    risk_level: str = Field(..., description="Risk level (Very Low, Low, Moderate, High, Very High)")
    score: float = Field(..., description="Risk score (0-100)")
    component_scores: Dict[str, float] = Field(..., description="Individual component scores")
    calculated_at: datetime

    class Config:
        json_schema_extra = {
            "example": {
                "ticker": "AAPL",
                "risk_level": "Low",
                "score": 25.5,
                "component_scores": {
                    "earnings_volatility": 20.0,
                    "financial_leverage": 15.0,
                    "operating_leverage": 30.0,
                    "concentration": 25.0,
                    "industry_stability": 35.0
                },
                "calculated_at": "2024-01-15T10:30:00Z"
            }
        }


class ManagementScoreResponse(BaseModel):
    """Management Score response schema."""
    ticker: str
    grade: str = Field(..., description="Management grade (A, B, C, D, F)")
    composite_score: float = Field(..., description="Composite management score")
    component_scores: Dict[str, float] = Field(..., description="Individual component scores")
    calculated_at: datetime

    class Config:
        json_schema_extra = {
            "example": {
                "ticker": "AAPL",
                "grade": "A",
                "composite_score": 85.5,
                "component_scores": {
                    "m_and_a": 80.0,
                    "capital_discipline": 90.0,
                    "shareholder_returns": 85.0,
                    "leverage_management": 88.0,
                    "governance": 84.0
                },
                "calculated_at": "2024-01-15T10:30:00Z"
            }
        }


class CompetitiveStrengthResponse(BaseModel):
    """Competitive Strength rating response schema."""
    ticker: str
    rating: str = Field(..., description="Rating (Dominant, Strong, Moderate, Weak, None)")
    composite_score: float = Field(..., description="Composite competitive strength score")
    component_scores: Dict[str, float] = Field(..., description="Individual component scores")
    trajectory: str = Field(..., description="Trajectory (Improving, Stable, Declining)")
    calculated_at: datetime

    class Config:
        json_schema_extra = {
            "example": {
                "ticker": "AAPL",
                "rating": "Dominant",
                "composite_score": 92.5,
                "component_scores": {
                    "network_effects": 95.0,
                    "intangible_assets": 90.0,
                    "cost_advantages": 88.0,
                    "switching_costs": 95.0,
                    "efficient_scale": 94.0
                },
                "trajectory": "Stable",
                "calculated_at": "2024-01-15T10:30:00Z"
            }
        }


class TSScoreResponse(BaseModel):
    """TradeSignal Score response schema."""
    ticker: str
    ts_score: float = Field(..., description="TradeSignal Score (0-100)")
    badge: str = Field(..., description="Score badge (Excellent, Strong, Good, Moderate, Weak, Poor)")
    components: Dict[str, float] = Field(..., description="Component scores")
    calculated_at: datetime

    class Config:
        json_schema_extra = {
            "example": {
                "ticker": "AAPL",
                "ts_score": 85.5,
                "badge": "Excellent",
                "components": {
                    "insider_buy_ratio": 0.75,
                    "ivt_discount_premium": 5.88,
                    "risk_score": 25.5,
                    "politician_score": 0.80
                },
                "calculated_at": "2024-01-15T10:30:00Z"
            }
        }


class ComprehensiveResearchResponse(BaseModel):
    """Comprehensive research data response schema."""
    ticker: str
    ivt: Optional[Dict[str, Optional[float]]] = Field(None, description="Intrinsic Value Target data")
    risk_level: Optional[Dict[str, Any]] = Field(None, description="Risk level assessment")
    ts_score: Optional[Dict[str, Any]] = Field(None, description="TradeSignal Score")
    management_score: Optional[Dict[str, Any]] = Field(None, description="Management score (Enterprise only)")
    competitive_strength: Optional[Dict[str, Any]] = Field(None, description="Competitive strength (Enterprise only)")

    class Config:
        json_schema_extra = {
            "example": {
                "ticker": "AAPL",
                "ivt": {
                    "intrinsic_value": 185.50,
                    "current_price": 175.20,
                    "discount_premium_pct": 5.88
                },
                "risk_level": {
                    "risk_level": "Low",
                    "score": 25.5
                },
                "ts_score": {
                    "score": 85.5,
                    "badge": "Excellent"
                },
                "management_score": {
                    "grade": "A",
                    "composite_score": 85.5
                },
                "competitive_strength": {
                    "rating": "Dominant",
                    "composite_score": 92.5,
                    "trajectory": "Stable"
                }
            }
        }

File: frontend/src/api/research.ts
import apiClient from './client';

export interface IVTData {
  ticker: string;
  intrinsic_value: number;
  current_price: number;
  discount_pct: number;
  calculation_date: string;
  confidence_score?: number;
  wacc?: number;
  terminal_growth_rate?: number;
  cached: boolean;
}

export interface TSScoreData {
  ticker: string;
  score: number;
  rating: string;
  price_to_ivt_ratio?: number;
  risk_adjusted: boolean;
  calculation_date?: string;
  cached: boolean;
}

export interface RiskLevelData {
  ticker: string;
  level: string;
  category: string;
  volatility_score?: number;
  assessment_date?: string;
  components: {
    earnings_volatility?: number;
    financial_leverage?: number;
    operating_leverage?: number;
    concentration_risk?: number;
    industry_stability?: number;
  };
  cached: boolean;
}

export interface CompetitiveStrengthData {
  ticker: string;
  rating: string;
  trajectory: string;
  composite_score?: number;
  moat_sources: {
    network_effects?: number;
    intangible_assets?: number;
    cost_advantages?: number;
    switching_costs?: number;
    efficient_scale?: number;
  };
  rating_date?: string;
  cached: boolean;
}

export interface ManagementScoreData {
  ticker: string;
  grade: string;
  composite_score?: number;
  components: {
    ma_track_record?: number;
    capital_discipline?: number;
    shareholder_returns?: number;
    financial_leverage?: number;
    governance?: number;
  };
  scoring_date?: string;
  cached: boolean;
}

export interface FullResearchReport {
  ticker: string;
  report_date?: string;
  ivt?: IVTData;
  ts_score?: TSScoreData;
  risk_level?: RiskLevelData;
  competitive_strength?: CompetitiveStrengthData;
  management_score?: ManagementScoreData;
}

export const researchApi = {
  /**
   * Get Intrinsic Value Target (IVT) for a stock
   * Returns null if data is not available (404) - this is expected for companies without coverage
   */
  getIVT: async (ticker: string): Promise<IVTData | null> => {
    try {
      const response = await apiClient.get(`/api/research/${ticker}/ivt`);
      return response.data;
    } catch (error: any) {
      // Silently return null for 404s (expected when data not available)
      if (error.response?.status === 404) {
        return null;
      }
      // Re-throw other errors
      throw error;
    }
  },

  /**
   * Get TradeSignal Score (TS Score) for a stock
   * Returns null if data is not available (404) - this is expected for companies without coverage
   */
  getTSScore: async (ticker: string): Promise<TSScoreData | null> => {
    try {
      const response = await apiClient.get(`/api/research/${ticker}/ts-score`);
      return response.data;
    } catch (error: any) {
      // Silently return null for 404s (expected when data not available)
      if (error.response?.status === 404) {
        return null;
      }
      // Re-throw other errors
      throw error;
    }
  },

  /**
   * Get Risk Level Assessment for a stock
   * Returns null if data is not available (404) - this is expected for companies without coverage
   */
  getRiskLevel: async (ticker: string): Promise<RiskLevelData | null> => {
    try {
      const response = await apiClient.get(`/api/research/${ticker}/risk-level`);
      return response.data;
    } catch (error: any) {
      // Silently return null for 404s (expected when data not available)
      if (error.response?.status === 404) {
        return null;
      }
      // Re-throw other errors
      throw error;
    }
  },

  /**
   * Get Competitive Strength Rating for a stock
   * Returns null if data is not available (404) - this is expected for companies without coverage
   */
  getCompetitiveStrength: async (ticker: string): Promise<CompetitiveStrengthData | null> => {
    try {
      const response = await apiClient.get(`/api/research/${ticker}/competitive-strength`);
      return response.data;
    } catch (error: any) {
      // Silently return null for 404s (expected when data not available)
      if (error.response?.status === 404) {
        return null;
      }
      // Re-throw other errors
      throw error;
    }
  },

  /**
   * Get Management Excellence Score for a stock
   * Returns null if data is not available (404) - this is expected for companies without coverage
   */
  getManagementScore: async (ticker: string): Promise<ManagementScoreData | null> => {
    try {
      const response = await apiClient.get(`/api/research/${ticker}/management-score`);
      return response.data;
    } catch (error: any) {
      // Silently return null for 404s (expected when data not available)
      if (error.response?.status === 404) {
        return null;
      }
      // Re-throw other errors
      throw error;
    }
  },

  /**
   * Get full research report with all scores
   * Returns null if data is not available (404) - this is expected for companies without coverage
   */
  getFullReport: async (ticker: string): Promise<FullResearchReport | null> => {
    try {
      const response = await apiClient.get(`/api/research/${ticker}/full-report`);
      return response.data;
    } catch (error: any) {
      // Silently return null for 404s (expected when data not available)
      if (error.response?.status === 404) {
        return null;
      }
      // Re-throw other errors
      throw error;
    }
  },

  /**
   * Get list of tickers with research coverage
   */
  getCoverage: async (): Promise<{
    total_coverage: number;
    tickers: string[];
    coverage_types: {
      ivt: number;
      ts_score: number;
    };
  }> => {
    const response = await apiClient.get('/api/research/coverage');
    return response.data;
  },
};

export default researchApi;
